### PR TITLE
Use ahash for BUILTIN_INSTRUCTION_COSTS

### LINKED ISSUE #2196: BUILT_IN_INSTRUCTION_COSTS: ahash
### Problem
- `BUILT_IN_INSTRUCTION_COSTS` using default hashing function
- Default hash function is slow

### Solution
- make `BUILT_IN_INSTRUCTION_COSTS` use `ahash`

### PR DESCRIPTION
**Pull request by ksolana**
Fixes: #2196

Bench results:

```
$ ./cargo nightly bench --manifest-path cost-model/Cargo.toml
```

### CODE REVIEW

**Comment by tao-stones** [comment]
Looks good, do you have bench results to share?

File: cost-model/benches/block_cost_limit.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        (ed25519_program::id(), 0),
+    ]
+    .iter()
+    .cloned()
+    .collect();
+    }
+
+    lazy_static! {
+    /// Number of compute units for each built-in programs
+    pub static ref hash_map: AHashMap<Pubkey, u64> = [
```
Comment (tao-stones) [inline cost-model/benches/block_cost_limit.rs]:
```suggestion
    pub static ref hash_map: HashMap<Pubkey, u64> = [
```

Comment (tao-stones) [reply cost-model/benches/block_cost_limit.rs]: To bench `BUILTIN_INSTRUCTION_COSTS`, you probably don't need to declare `ahash_map` nor `hash_map` here, instead just need one `bench_find()` that work against `BUILTIN_INSTRUCTION_COSTS`. So you run the bench when `BUILTIN_INSTRUCTION_COSTS` is hashmap, then run again after changingit to ahashmap, so we can compare the perf gain of the change.

**Comment by ksolana** [comment]
<details>
<summary>Benchmark Results</summary>

Baseline
```
test bench_hash_find ... bench:      30,955 ns/iter (+/- 756)
test bench_cost_model                       ... bench:   1,137,787 ns/iter (+/- 22,438)
test bench_cost_model_requested_write_locks ... bench:   1,174,591 ns/iter (+/- 12,127)
test bench_cost_tracker_contentious_transaction     ... bench:   1,831,908 ns/iter (+/- 122,044)
test bench_cost_tracker_non_contentious_transaction ... bench:   2,238,966 ns/iter (+/- 18,392)
test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out; finished in 2.60s
```

With changes
```
test bench_hash_find ... bench:      19,944 ns/iter (+/- 656)
test bench_cost_model                       ... bench:     817,650 ns/iter (+/- 25,232)
test bench_cost_model_requested_write_locks ... bench:     885,115 ns/iter (+/- 6,901)
test bench_cost_tracker_contentious_transaction     ... bench:   1,831,225 ns/iter (+/- 32,081)
test bench_cost_tracker_non_contentious_transaction ... bench:   2,246,248 ns/iter (+/- 113,709)
```

</details>

File: cost-model/Cargo.toml [hunk 2]
```diff
@@ -26,6 +26,7 @@ solana-sdk = { workspace = true }
 solana-stake-program = { workspace = true }
 solana-system-program = { workspace = true }
 solana-vote-program = { workspace = true }
+rand = "0.8.5"
```
Comment (tao-stones) [inline cost-model/Cargo.toml]: Not 100% sure with bench if this can be moved to `[dev-dependencies]`?

**Review by tao-stones — COMMENTED** [review]
thanks for adding bench, and the results look promising

**Review by tao-stones — APPROVED** [review]
lgtm

**Pull request merged**